### PR TITLE
Support adding untrusted root CA certificates (corp certs)

### PR DESCRIPTION
**Pull request by laozc**
This PR allows users to populate Corp Root CA certificates by adding certificates in PEM format to `~/.mininkube/certs`.
These certificates will be added to system certificate store, which will be trusted by all processes running in the minikube VM.

Closes: #1408

### CODE REVIEW

**Comment by RA489** [comment]
@minikube-bot OK to test

**Comment by medyagh** [comment]
@laozc Thank you so much for this PR ! this is actually very exciting !

do you mind sharing if you have tested this for yourself?

and I am very curious to find out how I could test it myself?

is there a way I could simulate a corp environment root CA and first see it fail and then after this PR see it succeed ?

**Comment by laozc** [comment]
I've verified the PR on both Linux and Windows (Hyper-V) myself.

I used the PR against my own CA inside a private network (mynetwork.net), which runs a private Docker image server (docker.mynetwork.net) with TLS enabled.

1) `curl MINIKUBE_IP:8443` (both in the VM and within the containers)

2) `curl https://mynetwork.net` (website which has an in-house corporate CA as the Root CA) 

3) `docker pull docker.mynetwork.net/ubuntu` (Note: the image server needs to install a SSL cert which has an in-house corporate CA as the Root CA)

There operations should not produce any errors like
`curl: (60) SSL certificate problem: self signed certificate in certificate chain`
or
`Error response from daemon: Get https://docker.mynetwork.net/v2/: x509: certificate signed by unknown authority`

You may use a self-signed root CA, generate a server cert with the root and OpenSSL to verify this PR.

File: pkg/minikube/bootstrapper/certs.go [hunk 1]
```diff
@@ -197,3 +218,117 @@ func generateCerts(k8s config.KubernetesConfig) error {
 
 	return nil
 }
+
+func collectCACerts() (map[string]string, error) {
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: Please add a comment here - describing where the certs are collected from.

File: pkg/minikube/bootstrapper/certs.go [hunk 2]
```diff
@@ -197,3 +218,117 @@ func generateCerts(k8s config.KubernetesConfig) error {
...
+	localPath := constants.GetMinipath()
+
+	isValidPem := func(hostpath string) (bool, error) {
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: Do you mind moving this function out from collectCACerts? With nested functions it is more difficult to understand scope.

File: pkg/minikube/bootstrapper/certs.go [hunk 3]
```diff
@@ -197,3 +218,117 @@ func generateCerts(k8s config.KubernetesConfig) error {
...
+	filtered := map[string]string{}
+	for k, v := range certFiles {
+		if v != "" {
+			filtered[k] = v
+		}
+	}
+	return filtered, nil
+}
+
+func configureCACerts(cmd command.Runner, caCerts map[string]string) error {
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: Add a comment to describe what this is configuring.

File: pkg/minikube/bootstrapper/certs.go [hunk 4]
```diff
@@ -197,3 +218,117 @@ func generateCerts(k8s config.KubernetesConfig) error {
...
+	getSubjectHash := func(hostpath string) (string, error) {
+		out, err := cmd.CombinedOutput(fmt.Sprintf("openssl x509 -hash -noout -in '%s'", hostpath))
+		if err != nil {
+			return "", err
+		}
+
+		stringHash := strings.ReplaceAll(out, "\n", "")
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: Would it be possible to use `strings.TrimSpace()`?

File: pkg/minikube/bootstrapper/certs.go [hunk 5]
```diff
@@ -197,3 +218,117 @@ func generateCerts(k8s config.KubernetesConfig) error {
...
+	for k, v := range certFiles {
+		if v != "" {
+			filtered[k] = v
+		}
+	}
+	return filtered, nil
+}
+
+func configureCACerts(cmd command.Runner, caCerts map[string]string) error {
+	getSubjectHash := func(hostpath string) (string, error) {
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: Move this function out from within configureCACerts as well.

File: pkg/minikube/constants/constants.go [hunk 6]
```diff
@@ -426,3 +426,8 @@ const (
 	// DriverDocumentation the documentation of the KVM driver
 	DriverDocumentation = "https://github.com/kubernetes/minikube/blob/master/docs/drivers.md"
 )
+
+const (
+	CACertificatesDir = "/usr/share/ca-certificates"
+	SSLCertStoreDir   = "/etc/ssl/certs"
```
Comment (tstromberg) [inline pkg/minikube/constants/constants.go]: Could these be moved to `bootstrapper/certs.go`? I prefer to keep constants as close to their user as possible, unless they must be shared by multiple packages.

File: pkg/minikube/bootstrapper/certs.go (hunk 2 above)
Comment (laozc) [reply pkg/minikube/bootstrapper/certs.go]: No problem. I'll move it out.

File: pkg/minikube/bootstrapper/certs.go (hunk 1 above)
Comment (laozc) [reply pkg/minikube/bootstrapper/certs.go]: Sure.

File: pkg/minikube/bootstrapper/certs.go (hunk 4 above)
Comment (laozc) [reply pkg/minikube/bootstrapper/certs.go]: Will use `strings.TrimSpace()` instead.

File: pkg/minikube/bootstrapper/certs.go [hunk 7]
```diff
@@ -38,6 +38,11 @@ import (
 	"k8s.io/minikube/pkg/util"
 )
 
+const (
+	CACertificatesDir = "/usr/share/ca-certificates"
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/certs.go]: is this only gonna work on linux ?

**Comment by medyagh** [comment]
> I've verified the PR on both Linux and Windows (Hyper-V) myself.
> 
> I used the PR against my own CA inside a private network (mynetwork.net), which runs a private Docker image server (docker.mynetwork.net) with TLS enabled.
> 
> 1. `curl MINIKUBE_IP:8443` (both in the VM and within the containers)
> 2. `curl https://mynetwork.net` (website which has an in-house corporate CA as the Root CA)
> 3. `docker pull docker.mynetwork.net/ubuntu` (Note: the image server needs to install a SSL cert which has an in-house corporate CA as the Root CA)
> 
> There operations should not produce any errors like
> `curl: (60) SSL certificate problem: self signed certificate in certificate chain`
> or
> `Error response from daemon: Get https://docker.mynetwork.net/v2/: x509: certificate signed by unknown authority`
> 
> You may use a self-signed root CA, generate a server cert with the root and OpenSSL to verify this PR.

I would love to see an integration tests for this, do you have scipts that I could run to exactly simulate those steps.

File: pkg/minikube/bootstrapper/certs.go (hunk 7 above)
Comment (laozc) [reply pkg/minikube/bootstrapper/certs.go]: The `openssl` command runs on Linux image so the host could be either Windows, Linux or Mac.

**Comment by RA489** [comment]
@minikube-bot OK to test.

**Comment by laozc** [comment]
Unfortunately except for generating SSL certificates, I don't have the script to launch a image repository with SSL certificate.

**Comment by medyagh** [comment]
> t for generating SSL certificates, I don't have the script to launch a image repository with SSL certificate.

could you add a tutorial on the cert maybe in the site?

like if I have a corp laptop with certs, which files do I have to copy and to where, exactly ?

btw this needs a rebase. I am excited to see this PR !

**Comment by medyagh** [comment]
here is an example of tutorials file:

https://github.com/kubernetes/minikube/blob/master/site/content/en/docs/Tutorials/continuous_integration.md

File: pkg/minikube/bootstrapper/certs.go [hunk 8]
```diff
@@ -197,3 +223,122 @@ func generateCerts(k8s config.KubernetesConfig) error {
...
+			return true, nil
+		}
+		fileBytes = rest
+	}
+
+	return false, nil
+}
+
+// collectCACerts looks up all PEM certificates with .crt or .pem extension in ~/.minikube/certs to copy to the host.
+// Minikube root CA is also included but libmachine certificates (ca.pem/cert.pem) are excluded.
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/certs.go]: minikube should always be lowercase:) after all we are mini !

File: pkg/minikube/bootstrapper/certs.go [hunk 9]
```diff
@@ -76,7 +97,7 @@ func SetupCerts(cmd command.Runner, k8s config.KubernetesConfig) error {
 	}
 
 	kubeCfg := api.NewConfig()
-	err := util.PopulateKubeConfig(kubeCfgSetup, kubeCfg)
+	err = util.PopulateKubeConfig(kubeCfgSetup, kubeCfg)
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/certs.go]: this was renamed to  kubeconfig.PopulateFromSettings

https://github.com/kubernetes/minikube/blob/master/pkg/minikube/kubeconfig/settings.go#L67

File: pkg/minikube/bootstrapper/certs.go [hunk 10]
```diff
@@ -66,6 +74,19 @@ func SetupCerts(cmd command.Runner, k8s config.KubernetesConfig) error {
...
+	for src, dst := range caCerts {
+		certFile, err := assets.NewFileAsset(src, path.Dir(dst), path.Base(dst), "0644")
+		if err != nil {
+			return err
+		}
+
+		copyableFiles = append(copyableFiles, certFile)
+	}
+
 	kubeCfgSetup := &util.KubeConfigSetup{
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/certs.go]: this was renamed to kubeconfig.Settings type Settings struct {

Comment (medyagh) [inline pkg/minikube/bootstrapper/certs.go]: this was refactored to name kubeconfig.Settings

**Comment by medyagh** [comment]
@laozc  This is beautiful work ! it looks good to be merged after the tests pass ! I feel impatient to merge it ! thank you for this !

**Comment by medyagh** [comment]
@laozc do you happen to know the exact error messages that minikube crashes on if they don't add the corp certs ? I think it would be nice to add a Solution message to the user and advice them to read this great tutorial you wrote.


BTW please feel free to add your name to the tutorial on the site as the author ! you deserve all the credit .

**Comment by laozc** [comment]
> @laozc do you happen to know the exact error messages that minikube crashes on if they don't add the corp certs ? I think it would be nice to add a Solution message to the user and advice them to read this great tutorial you wrote. something to be added here:
> 
> like this one: https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go#L97
> 

Let me reproduce it and make another change for a more user-friendly error prompt later.

> BTW please feel free to add your name to the tutorial on the site as the author ! you deserve all the credit .

Thanks. Happy to help.
Let me know if there is any concern.

**Pull request merged**